---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.path = "man/figures/README-",
    out.width = "100%"
)
```

# scUTRboot

<!-- badges: start -->
[![GitHub issues](https://img.shields.io/github/issues/mfansler/scutrboot)](https://github.com/mfansler/scutrboot/issues)
[![GitHub pulls](https://img.shields.io/github/issues-pr/mfansler/scutrboot)](https://github.com/mfansler/scutrboot/pulls)
<!-- badges: end -->

## Overview

The `scutrboot` package provides a collection of statistical tools for the analysis
of 3' UTR isoform usage from scRNA-seq data. In particular, it was developed to 
accompany the scUTRquant pipeline, working from 3' UTR isoform-level count data
to characterize isoform usage in groups of cells (e.g., clusters, cell types) and 
test differential usage across group. The "*boot*", both refers to the package's
role as the part of the scUTRx suite where the tools are kept and to the 
bootstrap-based procedures that are core to the statistical framework.

## Installation instructions

Get the latest stable `R` release from [CRAN](http://cran.r-project.org/). Then install `scutrboot` from [Bioconductor](http://bioconductor.org/) using the following code:

```{r 'install', eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}

# not available yet
# BiocManager::install("scutrboot")
```

And the development version from [GitHub](https://github.com/mfansler/scutrboot) with:

```{r 'install_dev', eval = FALSE}
BiocManager::install("mfansler/scutrboot")
```

## Usage

Application of the `scutrboot` package can be found in 
[the `scUTRquant-figures` repository](https://github.com/Mayrlab/scUTRquant-figures), 
particularly in the notebooks used to generate Figures 
[4](https://github.com/Mayrlab/scUTRquant-figures/tree/main/figures/figure4) and 
[5](https://github.com/Mayrlab/scUTRquant-figures/tree/main/figures/figure5).

## Citation

Below is the citation output from using `citation('scutrboot')` in R. Please
run this yourself to check for any updates on how to cite __scutrboot__.

```{r 'citation', eval = requireNamespace('scutrboot')}
print(citation('scutrboot'), bibtex = TRUE)
```

Please note that the `scutrboot` was only made possible thanks to many other R and bioinformatics software authors, which are cited either in the vignettes and/or the paper(s) describing this package.

## Code of Conduct

Please note that the `scutrboot` project is released with a [Contributor Code of Conduct](http://bioconductor.org/about/code-of-conduct/). By contributing to this project, you agree to abide by its terms.

## Development tools

* Continuous code testing is possible thanks to [GitHub actions](https://www.tidyverse.org/blog/2020/04/usethis-1-6-0/)  through `r BiocStyle::CRANpkg('usethis')`, `r BiocStyle::CRANpkg('remotes')`, and `r BiocStyle::CRANpkg('rcmdcheck')` customized to use [Bioconductor's docker containers](https://www.bioconductor.org/help/docker/) and `r BiocStyle::Biocpkg('BiocCheck')`.
* Code coverage assessment is possible thanks to [codecov](https://codecov.io/gh) and `r BiocStyle::CRANpkg('covr')`.
* The [documentation website](http://mfansler.github.io/scutrboot) is automatically updated thanks to `r BiocStyle::CRANpkg('pkgdown')`.
* The code is styled automatically thanks to `r BiocStyle::CRANpkg('styler')`.
* The documentation is formatted thanks to `r BiocStyle::CRANpkg('devtools')` and `r BiocStyle::CRANpkg('roxygen2')`.

This package was developed using `r BiocStyle::Biocpkg('biocthis')`.


